qir/qat/Passes/GroupingPass/Tests/Unit/grouping.cpp

```diff
@@ -0,0 +1,525 @@
+    }
+
+    return ir_manip;
+}
+} // namespace
+
+std::shared_ptr<ConfigurableQirAdaptorFactory> newQirAdaptor(ConfigurationManager& configuration_manager)
+{
+
+    auto adaptor = std::make_shared<ConfigurableQirAdaptorFactory>(configuration_manager);
+
+    configuration_manager.setConfig(TargetQisMappingPassConfiguration::createDisabled());
+    configuration_manager.setConfig(TargetQisConfiguration::fromQirTargetName("default"));
+    configuration_manager.setConfig(LlvmPassesConfiguration::createDisabled());
+
+    configuration_manager.setConfig(StaticResourceComponentConfiguration::createDisabled());
+    configuration_manager.setConfig(PostTransformValidationPassConfiguration::createDisabled());
+
+    return adaptor;
+}
+
+TEST(GroupingComponent, BasicGrouping)
+{
+    ConfigurationManager configuration_manager;
+    auto                 ir_manip = newIrManip(R"script(
+  %z = call i64 @get_arg()
+  %0 = icmp slt i64 %z, 0
+  %a1.i.i = select i1 %0, %Qubit* null, %Qubit* inttoptr (i64 1 to %Qubit*)
+  %a2.i.i = select i1 %0, %Qubit* inttoptr (i64 1 to %Qubit*), %Qubit* null
+  tail call void @__quantum__qis__x__body(%Qubit* %a1.i.i)
+  tail call void @__quantum__qis__z__body(%Qubit* null)
+  tail call void @__quantum__qis__cnot__body(%Qubit* %a1.i.i, %Qubit* %a2.i.i)
+  tail call void @__quantum__qis__mz__body(%Qubit* %a2.i.i, %Result* null)
+  tail call void @__quantum__qis__reset__body(%Qubit* %a2.i.i)
+  %1 = tail call i1 @__quantum__qis__read_result__body(%Result* null)
+  %2 = mul i64 %z, 45
+  %.op.op.i.i = add i64 %2, 13
+  %3 = select i1 %1, i64 -7, i64 %.op.op.i.i
+  call void @no_opt(i64 %3)
+  )script");
+
+    auto adaptor = newQirAdaptor(configuration_manager);
+    ir_manip->applyQirAdaptor(adaptor);
+
+    EXPECT_TRUE(ir_manip->hasInstructionSequence(
+        {
+            "%0 = call i64 @get_arg()",
+            "%1 = icmp slt i64 %0, 0",
+            "%2 = mul i64 %0, 45",
+            "%3 = add i64 %2, 13",
+            "br label %load",
+        },
+        "entry"));
+
+    EXPECT_TRUE(ir_manip->hasInstructionSequence(
+        {
+            "%4 = select i1 %1, %Qubit* null, %Qubit* inttoptr (i64 1 to %Qubit*)",
+            "%5 = select i1 %1, %Qubit* inttoptr (i64 1 to %Qubit*), %Qubit* null",
+            "br label %quantum",
+        },
+        "load"));
+
+    EXPECT_TRUE(ir_manip->hasInstructionSequence(
+        {
+            "tail call void @__quantum__qis__x__body(%Qubit* %4)",
+            "tail call void @__quantum__qis__z__body(%Qubit* null)",
+            "tail call void @__quantum__qis__cnot__body(%Qubit* %4, %Qubit* %5)",
+            "br label %readout",
+        },
+        "quantum"));
+
+    EXPECT_TRUE(ir_manip->hasInstructionSequence(
+        {
+            "tail call void @__quantum__qis__mz__body(%Qubit* %5, %Result* null)",
+            "tail call void @__quantum__qis__reset__body(%Qubit* %5)",
+            "%6 = tail call i1 @__quantum__qis__read_result__body(%Result* null)",
```

Comment:
I am a bit on the fence regarding what to do with read result (because I'd expect it to execute on the classical processor, whereas the measurements execute primarily on the quantum processor). What's your thinking for bundling it here, instead of with post-classical? I am good either way - another thing that I think we will see better what works well with more experience.